tutorials/transfer_learning/README.md

```diff
@@ -0,0 +1,254 @@
+
+Self contained Julia code presented in this tutorial is found in ["transfer_learning.jl"](transfer_learning.jl) and can be launched with: 
+
+```
+julia project=@. --threads=8 transfer_learning.jl
+```
+
+## Getting started
+
+In this tutorial, we'll used a pre-trained ResNet18 model to solve a 3-class classification problem: 🐱, 🐶, 🐼.
+
+Data can be accessed from [Kaggle](https://www.kaggle.com/datasets/ashishsaxena2209/animal-image-datasetdog-cat-and-panda).
+
+Following download and unzip, data is expected to live under the following structure: 
+
+```
+- data
+    - animals
+        - cats
+        - dogs
+        - panda
+```
+
+In Julia, the following packages are needed:
+
+```julia
+using Random: shuffle!
+import Base: length, getindex
+using Images
+using Flux
+using Flux: update!
+using DataAugmentation
+using Metalhead
+```
+
+We also define a utility to help manage cpu-gpu conversion:
+
+```julia
+device = Flux.CUDA.functional() ? gpu : cpu
+```
+
+A CUDA enabled GPU is recommended. A modest one with 6GB RAM is sufficient and should run the tutorial in just 5-6 mins. If running on CPU, it may take over 40 mins. 
+
+## Custom DataLoader
+
+When dealing with large datasets, it's unrealistic to use a vanilla `DataLoader` constructor using the entire dataset as input. A handy approach is to rely on custom data containers, which allows to only pull data into memory as needed. 
+
+Our custom data container is very simple. It's a `struct` containing the paths to each of our images: 
+
+```julia
+const CATS = readdir(abspath(joinpath("data", "animals", "cats")), join = true)
+const DOGS = readdir(abspath(joinpath("data", "animals", "dogs")), join = true)
+const PANDA = readdir(abspath(joinpath("data", "animals", "panda")), join = true)
+
+struct ImageContainer{T<:Vector}
+    img::T
+end
+
+imgs = [CATS..., DOGS..., PANDA...]
+shuffle!(imgs)
+data = ImageContainer(imgs)
+```
+
+In order to be compatible with `DataLoader`, 2 functions must minimally be defined:
+ - `Base.length`: returns the number of observations in the data container.
+ - `Base.getindex`: function that returns the observation for a specified index. 
+
+```julia
+length(data::ImageContainer) = length(data.img)
+
+const im_size = (224, 224)
+tfm = DataAugmentation.compose(ScaleKeepAspect(im_size), CenterCrop(im_size))
+name_to_idx = Dict{String,Int32}("cats" => 1, "dogs" => 2, "panda" => 3)
+
+function getindex(data::ImageContainer, idx::Int)
+    path = data.img[idx]
+    img = Images.load(path)
+    img = apply(tfm, Image(img))
+    img = permutedims(channelview(RGB.(itemdata(img))), (3, 2, 1))
+    img = Float32.(img)
+    name = replace(path, r"(.+)\\(.+)\\(.+_\d+)\.jpg" => s"\2")    
+    y = name_to_idx[name]
+    return img, Flux.onehotbatch(y, 1:3)
+end
```

Comment:
instead of applying the transformation in `getindex` it would be better to showcase `MLUtils.mapobs`. The advantage is that it is a pattern that can be used with any dataset. An example with mnist is given in the README [here](https://github.com/CarloLucibello/Tsunami.jl)

Reply:
Is what you suggest to have a sepration between a minimal `getindex` and a transform function, so something like: 

```julia
function getindex(data::ImageContainer, idx::Int)
    path = data.img[idx]
    x = Images.load(path)
    name = replace(path, r"(.+)\\(.+)\\(.+_\d+)\.jpg" => s"\2")
    y = name_to_idx[name]
    return (x = x, y = y)
end
function img_transform(batch)
    img = apply(tfm, Image(batch[:x]))
    img = permutedims(channelview(RGB.(itemdata(img))), (3, 2, 1))
    x = Float32.(img)
    y = Flux.onehotbatch(batch[:y], 1:3)
    return (x, y)
end

dtrain = Flux.mapobs(img_transform, ImageContainer(imgs[1:2700]))
``` 

That's out of scope for thie PR, but benchmarking to validate there wasn't any performance difference, both approaches perform essentially the same, but performance degrades following each iteration: 

```julia
function data_loop(data)
    count = 0
    for (x, y) in data
        count += size(y, 2)
    end
    @info count
end
@btime data_loop($dtrain)
2.778 s (505360 allocations: 6.86 GiB)
4.000 s (505459 allocations: 6.86 GiB)
5.303 s (505461 allocations: 6.86 GiB)
5.423 s (505424 allocations: 6.86 GiB)
```

That's with MLUtils 2.11, so may be ignored until tutorial can be updated to latest Flux/MLUtils versions.

Reply:
> Is what you suggest to have a sepration between a minimal getindex and a transform function

yes

>  but performance degrades following each iteration

this is awful, if you have some time please open an issue MLUtils.jl

For this PR do as you prefer regarding the use of mapobs or not